### ISSUE TITLE
"No connection could be made" error when running with OpenCover

### ISSUE THREAD

**Issue by Roemer** [nunit/nunit-console#239]
I get an exception after some tests with the error:
> No connection could be made because the target machine actively refused it 127.0.0.1:9548
Failed to stop the remote agent. No connection could be made because the target machine actively refused it 127.0.0.1:9548

When I run a specific set of tests with nunit inside OpenCover. When I run the tests with nunit alone they all run correctly.
On some CI slaves, the nunit-agent even crashes and it seems it gets into a StackOverflow exception. I even tried the version from the current CI with the same result.
I am not totally sure if it is an nunit problem (but I think that is where the connection refused comes from, probably because the nunit-agent crashes).

I'm a bit stuck now. Do you have any ideas how I could possibly debug that issue?
Thanks for your help.

**Comment by jnm2** [comment]
I've gotten this 5 times out of the last 9 builds on the build server at work, sometimes in the same build log as this error: https://github.com/nunit/nunit-console/issues/285 (Yup, so far, no builds are going through.) The only thing that should have changed from last night is that there's a lot more parallelization going on.

```
##[error]Unhandled Exception: NUnit.Engine.NUnitEngineException: No connection could be made because the target machine actively refused it 127.0.0.1:54259
##[error]Failed to stop the remote agent. No connection could be made because the target machine actively refused it 127.0.0.1:54259
##[error]   at NUnit.Engine.Runners.ProcessRunner.Dispose(Boolean disposing)
##[error]   at NUnit.Engine.Runners.AbstractTestRunner.Dispose()
##[error]   at NUnit.Engine.Runners.TestExecutionTask.Execute()
##[error]   at NUnit.Engine.Runners.ParallelTaskWorkerPool.ProcessTasksProc()
##[error]   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
##[error]   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
##[error]   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state)
##[error]   at System.Threading.ThreadHelper.ThreadStart()
```

**Comment by jnm2** [comment]
After running this on my dev machine, I'm super suspicious that the problem is really caused by my tests pegging the CPU. Too parallel perhaps. In my case NUnit doesn't know it but my Cake script is running three instances of the NUnit console runner, each with tests in parallel.

**Comment by jnm2** [comment]
I had to roll back to NUnit Framework 3.7.1 because of https://github.com/nunit/nunit-console/issues/285 and now this problem disappeared, too. So there goes that theory.

**Comment by jnm2** [comment]
Now that framework 3.8.1 is out, this error comes back when I roll forward from 3.7.1 to 3.8.1. (Fyi @rprouse)

**Comment by jnm2** [comment]
Shows up alongside the elusive SocketException!

**Comment by jnm2** [comment]
But... if I make all my tests non-parallelizable, the SocketException and this "target machine actively refused" problem go away! Parallelism seems to be key.

**Comment by jnm2** [comment]
Here's what I've learned while stabilizing our build server at work:

 - Running a single console runner process at a time, with tests parallelized by NUnit, no issues show up.

 - Running three console runner processes in parallel, with tests *not* parallelized by NUnit, no issues show up.

 - Running three console runner processes in parallel, with tests parallelized by NUnit, often causes "No connection could be made because the target machine actively refused it."
   But, it also causes the build server CPU to peg, LocalDB to timeout half the time, and when I'm running on my dev machine, my cursor to freeze. **Therefore I'm putting this down to a CPU starvation problem external to NUnit which NUnit has no control over.**

Running the three console runner processes in parallel was a good idea, but now that NUnit is parallelizing all tests inside each process, there's no benefit. It's just causing issues when running  3×(number of CPUs) power-hungry threads.

@Roemer do you think this is possibly similar to your situation?

**Comment by CharliePoole** [comment]
@jnm2 Since NUnit has two different ways to parallelize, your "tests parallelized by NUnit" is ambiguous. I'm assuming you mean parallel workers within the framework rather than multiple process agents, in each of the cases... right?

Everyone... NUnit provides multiple ways to parallelize things so you can pick one. In most situations, you shouldn't use all of them simultaneously without giving it some thought, as @jnm2 has done. 

Paralllel agents is essentially equivalent to running multiple console processes, except you get (1) automatic throttling to the number of agents you set and (2) combined XML output.

Parallel workers allows you to parallelize at the fixture or test case level within an assembly. You get automatic throttling based on the number of cores __but__ that's only useful if it's the only assembly being run. Unfortunately, I'm afraid, we run a separate process for each assembly by default. If each of those processes is trying to use all the cores up, you probably will have an issue. You can limit the number used for each assembly through the `LevelOfParallelism` attribute.

Parallel performance is complicated!

**Comment by jnm2** [comment]
@CharliePoole good point, clarified.

**Comment by jnm2** [comment]
Maybe NUnit could give all threads `ThreadPriority.BelowNormal` to avoid accidentally starting too many processes each of which can max each core.

**Comment by jnm2** [comment]
I hate having to retract comments, though it's the only way I can keep track of things, but I just hit this again for the first time in the first scenario (Running a single console runner process at a time, with tests `[Parallelizable]` via NUnit framework).

**Comment by jnm2** [comment]
Happened twice in a row.

**Comment by jnm2** [comment]
It's my unit tests causing this, not the integration tests. I guess because the IO slows them down.

**Comment by jnm2** [comment]
Happened five times in a row. I rolled back to 3.7.1 and it has succeeded once so far.

**Comment by jnm2** [comment]
Happened five times in a row. I rolled back to 3.7.1 and it has succeeded five times so far.

On the bright side, I seem to have a guaranteed repro now.

**Comment by Roemer** [comment]
For me the problem lies in using AutoMapper with OpenCover (and intensifies when using that in IIS). This often causes an infinite loop somewhere in AutoMapper which in the end leads to a StackOverflow which results in the error mentioned in this issue.

**Comment by jnm2** [comment]
@Roemer You're saying that this is just obscuring the message of the initial error? Interesting...

**Comment by jnm2** [comment]
I just hit this at work with NUnit 3.7.1 (and console 3.7.0), so staying away from 3.8.x isn't good enough for now.

**Comment by dpisanu** [comment]
I am running into the same issue here. Using **NUnit Console Runner 3.8.0**.
I have a collection of Test Assemblies ranging from 10 to 400 individual Test Stubs.
Some basic Unit Tests, some are more elaborate Component Tests up to complex UI Integration Tests using **TestStack.White** and **FlaUI** (@Roemer hey there *g*).

It's exactly when running the UI Tests that we tend to get the connection issue.
it seems it's a side effect of the Workstation getting locked (RDP connection Idles out for Example) and then a burried and unhandled exception in the running Test and Test Framework bubbles up and leads to the Test Agent to trip.

It's apparent because when I run **nunit-console** with the _--inprocess_ option, it also fumbles, just with a different error result.
Running in verbose mode does not provide much more input. the internalTrace logs don't show much.

**Comment by jnm2** [comment]
@dpisanu Do you think you could build something we could reliably repro on our machines?

**Comment by dpisanu** [comment]
I can try.
Don't have a demo right at hand so will need to chip some time off from the normal schedule.
Let me get back to you on that.

**Comment by jnm2** [comment]
That would be great!

> It's apparent because when I run nunit-console with the --inprocess option, it also fumbles, just with a different error result.

I'm interested to see what this is too.

**Comment by ChrisMaddock** [comment]
Closing this, as it's been nearly 3 years since the last update, and unfortunately there isn't enough information here to understand/reproduce the problem. 

Happy to reopen if anyone has more to add. 🙂

**Comment by chammonsvusion** [comment]
Hate to open this up again .. but I am having the "exact" issue.  Except it fails everytime .. with or without opencover.  I added about 20 unit tests which ran fine in Visual Studio but not on our Jenkins build server.

Here is simplest form of error .. with just offending DLL  without OpenCover ...

```
C:\ws\VusionPackages_master - Copy (2)>"C:\Program Files (x86)\NUnit.org\nunit-console\nunit3-console.exe" .\AntonyAndCleopatra.Common\Utility\Utilities.Test\bin\x64\Release\Utilities.Test.dll  --result=testresults-AntonyAndCleopatra.Common.xml
NUnit Console Runner 3.11.1 (.NET 2.0)
Copyright (c) 2020 Charlie Poole, Rob Prouse
Wednesday, June 23, 2021 7:47:36 PM

Runtime Environment
   OS Version: Microsoft Windows NT 6.2.9200.0
   Runtime: .NET Framework CLR v4.0.30319.42000

Test Files
    .\AntonyAndCleopatra.Common\Utility\Utilities.Test\bin\x64\Release\Utilities.Test.dll


Unhandled Exception: NUnit.Engine.NUnitEngineException: Remote test agent exited with non-zero exit code -1073741571
   at NUnit.Engine.Services.TestAgency.OnAgentExit(Process process, Guid agentId)
   at NUnit.Engine.Services.TestAgency.<>c__DisplayClass7_0.<LaunchAgentProcess>b__0(Object sender, EventArgs e)
   at System.Diagnostics.Process.OnExited()
   at System.Diagnostics.Process.RaiseOnExited()
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
   at System.Threading._ThreadPoolWaitOrTimerCallback.PerformWaitOrTimerCallback(Object state, Boolean timedOut)

Errors, Failures and Warnings

1) Error :
System.Net.Sockets.SocketException : An existing connection was forcibly closed by the remote host
--SocketException
An existing connection was forcibly closed by the remote host

Server stack trace:
   at System.Runtime.Remoting.Channels.SocketStream.Read(Byte[] buffer, Int32 offset, Int32 size)
   at System.Runtime.Remoting.Channels.SocketHandler.ReadFromSocket(Byte[] buffer, Int32 offset, Int32 count)
   at System.Runtime.Remoting.Channels.SocketHandler.Read(Byte[] buffer, Int32 offset, Int32 count)
   at System.Runtime.Remoting.Channels.SocketHandler.ReadAndMatchFourBytes(Byte[] buffer)
   at System.Runtime.Remoting.Channels.Tcp.TcpSocketHandler.ReadAndMatchPreamble()
   at System.Runtime.Remoting.Channels.Tcp.TcpSocketHandler.ReadVersionAndOperation(UInt16& operation)
   at System.Runtime.Remoting.Channels.Tcp.TcpClientSocketHandler.ReadHeaders()
   at System.Runtime.Remoting.Channels.Tcp.TcpClientTransportSink.ProcessMessage(IMessage msg, ITransportHeaders requestHeaders, Stream requestStream, ITransportHeaders& responseHeaders, Stream& responseStream)
   at System.Runtime.Remoting.Channels.BinaryClientFormatterSink.SyncProcessMessage(IMessage msg)

Exception rethrown at [0]:
   at System.Runtime.Remoting.Proxies.RealProxy.HandleReturnMessage(IMessage reqMsg, IMessage retMsg)
   at System.Runtime.Remoting.Proxies.RealProxy.PrivateInvoke(MessageData& msgData, Int32 type)
   at NUnit.Engine.ITestEngineRunner.Run(ITestEventListener listener, TestFilter filter)
   at NUnit.Engine.Runners.ProcessRunner.RunTests(ITestEventListener listener, TestFilter filter)

Test Run Summary
  Overall result: Failed
  Test Count: 0, Passed: 0, Failed: 0, Warnings: 0, Inconclusive: 0, Skipped: 0
  Start time: 2021-06-24 00:47:36Z
    End time: 2021-06-24 00:47:57Z
    Duration: 20.759 seconds

Results (nunit3) saved as testresults-AntonyAndCleopatra.Common.xml

NUnit.Engine.NUnitEngineUnloadException : Agent Process was terminated successfully after error.
  ----> System.Net.Sockets.SocketException : No connection could be made because the target machine actively refused it 127.0.0.1:57740

```

**Comment by mikkelbu** [comment]
@chammonsvusion This is not the same issue (except that the agent processes are terminated). Your stack trace includes `Unhandled Exception: NUnit.Engine.NUnitEngineException: Remote test agent exited with non-zero exit code -1073741571` which means that the process exists with a Stack Overflow exception - see e.g. https://stackoverflow.com/questions/20629027/process-finished-with-exit-code-1073741571.

To find the offending test you can try to run the tests in the same process using `--inprocess` or using `--labales=Before` to see which of the tests is triggering the stack overflow.

Ps. If you update the console to 3.12 then the console will tell you `Remote test agent was terminated due to a stack overflow`.

**Comment by CharliePoole** [comment]
Fun fact: In the earliest days of NUnit V2, running under .NET 1.1, we were able to display the stack trace for a stack overflow exception, at least up to the point where the stack ran out of memory. I wrote code to abbreviate the display by counting the number of duplicate entries. After the release of .NET 2.0, this was no longer possible.

Apparently, 1.1 kept a bit of memory in reserve, which allowed us to catch the exception and do this. I've been told that the 2.0 developers considered this "dangerous" since we might actually try to recover with corrupted memory, but it was definitely handy for our specific use.
